### PR TITLE
Updating Task Framework to support User Defined Workflow implementation

### PR DESCRIPTION
**Pull request by yasgun**
Changes to the existing Helix based Task Framework to support User Defined Workflow implementation

### CODE REVIEW

**Comment by DImuthuUpe** [comment]
Looks good @yasgun. This a general but important improvement for Task Framework. I'm just wondering why I'm seeing [DO NOT MERGE] as the title :)

File: modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * <p>
+ * Queue will be there until user delete it.
+ * Queue can keep accepting tasks.
+ * No parallel run allowed except intentionally configured.
+ */
+public class QueueOperator {
+
+    private final static Logger logger = LoggerFactory.getLogger(QueueOperator.class);
+
+    private static final String WORKFLOW_PREFIX = "Job_queue_";
```
Comment (DImuthuUpe) [inline modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java]: Rename it as Queue Prefix

File: modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        Runtime.getRuntime().addShutdownHook(
+                new Thread(() -> helixManager.disconnect())
+        );
+    }
+
+    public synchronized String createJobQueue(String queueName, boolean monitor) throws Exception {
+
+        String workflowName = WORKFLOW_PREFIX + queueName;
+        logger.info("Launching workflow " + workflowName + " for job queue " + queueName);
```
Comment (DImuthuUpe) [inline modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java]: Launching queue

File: modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        workflowCfgBuilder.setExpiry(0);
+
+        JobQueue.Builder jobQueueBuilder = new JobQueue.Builder(queueName).setWorkflowConfig(workflowCfgBuilder.build());
+        taskQueues.put(queueName, new TaskQueue(jobQueueBuilder, workflowName, queueName, monitor));
+
+        return workflowName;
+    }
+
+    public synchronized String addTaskToQueue(String queueName, AbstractTask task, boolean globalParticipant) throws IllegalAccessException {
+        return taskQueues.get(queueName).addTask(task, globalParticipant);
```
Comment (DImuthuUpe) [inline modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java]: print logs

File: modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    public synchronized String removeTaskFromQueue(String queueName, String taskId) throws InterruptedException {
+        return taskQueues.get(queueName).removeTask(queueName, taskId);
```
Comment (DImuthuUpe) [inline modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java]: print logs

File: modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+ * Queue can keep accepting tasks.
+ * No parallel run allowed except intentionally configured.
+ */
+public class QueueOperator {
+
+    private final static Logger logger = LoggerFactory.getLogger(QueueOperator.class);
+
+    private static final String WORKFLOW_PREFIX = "Job_queue_";
+    private HelixManager helixManager;
+    private HashMap<String, TaskQueue> taskQueues = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (DImuthuUpe) [reply modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java]: Operator should be stateless. You shouldn't keep application specific details at this level as they are not recoverable if gone. You can save these details on Zookeeper but by looking at this TaskQueue code, I think it is a utility class where you don't need to store for each queue.

**Comment by DImuthuUpe** [comment]
@yasgun This is a good start point for introducing task queuing capability into Task Framework. I have added few comments.

**Comment by yasgun** [comment]
@DImuthuUpe Thank you for the clarification. I updated the Queue Operator to become stateless.

File: modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        taskDriver.enqueueJob(queueName, task.getTaskId(), job);
+
+        return task.getTaskId();
+    }
+
+    public synchronized String removeTaskFromQueue(String queueName, String taskId) throws InterruptedException {
+        logger.info("Removing task: " + taskId + " from queue: " + queueName);
+        taskDriver.stop(queueName);
```
Comment (DImuthuUpe) [inline modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java]: What happens if the task is already executed?

File: modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        if (monitor) {
+            TaskState taskState = taskDriver.pollForWorkflowState(queueName,
+                    TaskState.COMPLETED, TaskState.FAILED, TaskState.STOPPED, TaskState.ABORTED);
+            logger.info("Queue " + queueName + " finished with state " + taskState.name());
+        }
+
+        return queueName;
+    }
+
+    public synchronized void stopQueue(String queueName) throws InterruptedException {
```
Comment (DImuthuUpe) [inline modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java]: Add java doc to public api methods and the assumptions that you have done (if any)

File: modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        logger.info("Stopping queue: " + queueName);
+        taskDriver.stop(queueName);
```
Comment (DImuthuUpe) [inline modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java]: Might have to check whether there is a queue with that name?

File: modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        logger.info("Resuming queue: " + queueName);
+        taskDriver.resume(queueName);
+    }
+
+    public synchronized void deleteQueue(String queueName) {
+        logger.info("Deleting queue: " + queueName);
+        taskDriver.delete(queueName);
+    }
+
+    public synchronized void cleanupQueue(String queueName) throws InterruptedException {
```
Comment (DImuthuUpe) [inline modules/airavata-helix/workflow-impl/src/main/java/org/apache/airavata/helix/workflow/QueueOperator.java]: What is the purpose of cleanup? Clearly mention in java doc.

**Comment by DImuthuUpe** [comment]
Looks really great @yasgun. Now code is clear and simple. I added few more comments specially related to java doc. You might have to do some validations as well. We cannot expect helix to do that.

File: modules/airavata-helix/task-core/src/main/java/org/apache/airavata/helix/core/participant/HelixParticipant.java [hunk 10]
```diff
@@ -60,30 +58,40 @@
...
+
     private String taskTypeName;
     private PropertyResolver propertyResolver;
-    private Class<T> taskClass;
+
+    private List<Class<? extends T>> taskClasses;
     private final List<AbstractTask> runningTasks = Collections.synchronizedList(new ArrayList<AbstractTask>());
 
-    public HelixParticipant(Class<T> taskClass, String taskTypeName) throws ApplicationSettingsException {
+    public HelixParticipant(List<Class<? extends T>> taskClasses, String taskTypeName) throws ApplicationSettingsException {
```
Comment (DImuthuUpe) [inline modules/airavata-helix/task-core/src/main/java/org/apache/airavata/helix/core/participant/HelixParticipant.java]: This is not relevant to this PR. However please enable the participant manually when it is registering in to the cluster. Participant become disabled when they are gracefully turned off and do not go into enabled state when they come back to the cluster

Comment (yasgun) [reply modules/airavata-helix/task-core/src/main/java/org/apache/airavata/helix/core/participant/HelixParticipant.java]: I have updated the PR with a modification to the HelixParticipant. @DImuthuUpe can you verify whether this addresses the actual requirement you have mentioned?

Comment (yasgun) [reply modules/airavata-helix/task-core/src/main/java/org/apache/airavata/helix/core/participant/HelixParticipant.java]: @DImuthuUpe I tested the implementation in my local environment and it is working as expected.

**Pull request merged**